File: bifurcate/cipher.py
"""The Feistel network — block-level encrypt/decrypt.

    Encrypt:  for i = 0..11:   L, R = R, L XOR F(R, K_i)   ; output R || L
    Decrypt:  identical, with the round keys consumed in reverse order.

Encryption and decryption are THE SAME FUNCTION. Only the key order flips.
That is the elegance of Feistel:
whatever F does — however non-invertible, however chaotic — the structure
still inverts.
"""

from __future__ import annotations

from .keyschedule import HALF_BYTES, ROUNDS, RoundKeys, expand
from .round import F

BLOCK_BYTES = 16  # 128-bit block


def _feistel(block: bytes, rk: RoundKeys, *, reverse: bool) -> bytes:
    if len(block) != BLOCK_BYTES:
        raise ValueError(f"block must be exactly {BLOCK_BYTES} bytes")
    left, right = block[:HALF_BYTES], block[HALF_BYTES:]
    order = range(ROUNDS - 1, -1, -1) if reverse else range(ROUNDS)
    for i in order:
        f = F(right, rk.keys[i], rk.rots[i], rk.sbox)
        left, right = right, bytes(a ^ b for a, b in zip(left, f))
    return right + left  # final swap omitted -> concatenate reversed


def encrypt_block(block: bytes, rk: RoundKeys) -> bytes:
    """Encrypt one 128-bit block."""
    return _feistel(block, rk, reverse=False)


def decrypt_block(block: bytes, rk: RoundKeys) -> bytes:
    """Decrypt one 128-bit block."""
    return _feistel(block, rk, reverse=True)


class Bifurcate:
    """Convenience wrapper binding a key schedule to the block operations."""

    def __init__(self, enc_seed: bytes) -> None:
        self.rk = expand(enc_seed)

    def encrypt_block(self, block: bytes) -> bytes:
        return encrypt_block(block, self.rk)

    def decrypt_block(self, block: bytes) -> bytes:
        return decrypt_block(block, self.rk)

File: cipher-ts/src/cipher.ts
// The Feistel network — bifurcate/cipher.py.
// Encryption and decryption are the SAME routine; only the key order flips.
import { RoundKeys, ROUNDS, expand } from "./keyschedule.js";
import { F, HALF_BYTES } from "./round.js";
import { concat, xorBytes } from "./bytes.js";

export const BLOCK_BYTES = 16;

function feistel(block: Uint8Array, rk: RoundKeys, reverse: boolean): Uint8Array {
  if (block.length !== BLOCK_BYTES) throw new Error("block must be 16 bytes");
  let left = block.slice(0, HALF_BYTES);
  let right = block.slice(HALF_BYTES);
  const order = reverse ? [...Array(ROUNDS).keys()].reverse() : [...Array(ROUNDS).keys()];
  for (const i of order) {
    const f = F(right, rk.keys[i]!, rk.rots[i]!, rk.sbox);
    const nr = xorBytes(left, f);
    left = right; right = nr;
  }
  return concat(right, left); // final swap: R || L
}

export const encryptBlock = (block: Uint8Array, rk: RoundKeys) => feistel(block, rk, false);
export const decryptBlock = (block: Uint8Array, rk: RoundKeys) => feistel(block, rk, true);

export class Bifurcate {
  rk: RoundKeys;
  constructor(encSeed: Uint8Array) { this.rk = expand(encSeed); }
  encryptBlock = (b: Uint8Array) => encryptBlock(b, this.rk);
  decryptBlock = (b: Uint8Array) => decryptBlock(b, this.rk);
}
